### ISSUE TITLE
Skip a test from the inside of the It script block

### ISSUE THREAD

**Issue by michalporeba** [pester/Pester#1022]
## 1. Provide a general summary of the issue in the Title above

My usecase is for environmental validation, not unit tests as such. I have been recently working on dbachecks project trying to find an efficient way of structuring code while making sure we can unit test environmental tests. One of the remaining challanges is that to keep code simple and any logic testable I don't want to depend on the `-Skip` parameter of `It`. 

Is there a way, or do you see it possible, to skip the test from inside the `It` script block similarily to how you can now mark the test as inconclusive? 

## 3. Expected Behavior

```
@($testCases).ForEach{
  It "My Test" {
     if ($psitem.MyProperty1 -eq $true) { Set-TestSkipped } else {
        $psitem.MyProperty2 | Should -Be 42 -Bdcause "it is the reason behind everything"
     }
  }
}
```

or similarily using test cases 
```
It "My Test" -TestCases $testCases {
   param($myProperty1, $myProperty2)
   if ($myProperty1 -eq $true) { Set-TestSkipped } else {
      $myProperty2 | Should -Be 42 -Bdcause "it is the reason behind everything"
   }
}
```

I'd rather not use the Inconlusive output, as to me it is very different status. Inconclusive is a test which cannot be asserted at a time for some reason, skipped is something that shouldn't even be asserted.

**Comment by nohwnd** [comment]
@michalporeba That is not possible for the moment, I was not even very happy about addition of the set-testinconclusive function, it's not a very systematic solution. It would be much nicer to have access to the context of the test from the insides of the test via some object that we could evolve.

**Comment by michalporeba** [comment]
@nohwnd Could you point me where to start looking at it to see how difficult it would be to make it possible in the future? I agree that having access to some 'context' would be much better than a global 'Set-TestInconclusive' command.

**Comment by nohwnd** [comment]
@michalporeba The `Set-TestInconclusive` function throws an exception with special error id. This error is recognized by `ConvertTo-PesterResult` and produces inconclusive result. If I was implementing the functionality you want I would add `Get-Test` (or Get-PesterTest?) that would look at the current test info, and return a live read-only object with the test name, params etc. Then I would have `Set-Test -State` that would probably throw as well, and the info would be parsed as part of the failing test. 

Implementing should be quite simple (look in It.ps1, PesterState.ps1 and Output.ps1), but I am not sure how the functionality is. Do you have some other usages in mind other than failing the test? What we could use reading the test state for?

**Comment by michalporeba** [comment]
@nohwnd Thank you, I'll have a look at the code in the next few days (hopefully). Main main usage cases are around environmental checks. Please have a look at what I have written about it here (http://dbain.wales/about/environmental-checks). I would like to be able to skip the test from inside the It block so that the skipping logic can be tested. Potentially some improvements could be made to make the failing check testing a bit better, but I'll have to think about it a bit more. Perhaps ability to pass the reason for skipping the test? 

I will review the code and report back here what I think. For now I was thinkin that perhaps having the It script block return some value could be a solution? I don't know, at this moment I'm just guessing.

**Comment by nohwnd** [comment]
You can attach data to the exception and read them from it afterwards.

**Comment by michalporeba** [comment]
The more I think about it the more I disagree with my initial instincts that the Set-TestInconclusive is ugly and a bit awkward solution. Throwing exceptions might be the best way to implement it and a function that throws an exception with correctly initialized object attached to it might be the best, and most elegant solution. I'm thinking about it from the perspective of environmental checks and ability to unit test it without too much unnecessary verbiage in checks and unit tests files. I've got another idea for improvement. I'll describe it first in another issue and I will come back and explain more in this thread.

**Comment by michalporeba** [comment]
Please see #1025 and the referenced blog post. I think it is essential that the logic in the check's `It` block is encapsulated in a `Confirm-*` function so that it can be unit tested. That means in the simplest of examples we have an environmental check that looks like so
```PowerShell
It "Check I can ping something" {
   Confirm-CanPingSomething
}
```
That is simple enough, but in practice, with context and reason it already is
```PowerShell
$config = Get-ConfigForCanPingSomethingCheck
@(Get-MyTestEntities).ForEach{
  It "Check I can ping something" {
    $psitem | Confirm-CanPingSomething -With $config -Because "there is a reason we need it"
  }
}
```
It gets already a bit more complicated to understand, and in my opinion checks should be easily understood by sysadmins who don't necessarily are developers. Passing some context down would probably required defining parameters in the script block and then passing them to Confirm-CanPingSomething. Then if we wanted to avoid exception, we need to pass the results back and soon we end up with something like this
```PowerShell
$config = Get-ConfigForCanPingSomethingCheck
@(Get-MyTestEntities).ForEach{
  It "Check I can ping something" {
    param([ContextObject]$context)
    return ($psitem | Confirm-CanPingSomething -With $config -Context $context -Because "there is a reason we need it")
  }
}
```

The more I think about it the less I like it. Throwing an exception with some context object attached to it from inside `Confirm-CanPingSomething` and then catching it inside `It` would allow for much cleaner looking unit tests, and environmental checks.

**Comment by michalporeba** [comment]
@nohwnd if you have a moment please look here 
https://github.com/michalporeba/Pester/commit/87b0ee17146493c8b6269c94cc922004b62b758d
in this commit I have a first draft of how skipping could look like. Inconclusive is already included in the same way. Perhaps pending should be included too?

Let me know what you think

**Comment by michalporeba** [comment]
![image](https://user-images.githubusercontent.com/120880/38675405-7dab6766-3e4f-11e8-94a2-24f780725f02.png)

**Comment by nohwnd** [comment]
I like the inclusion of `-Because`, what will happen if I simply skip the test with the skip parameter? :) The command name I don't like much, I think it should rather be called `Set-TestResult` (or `Set-ItResult`).  Internally I would rather switch on the error Id to figure out the result, replacing parts of the error Id is clever, but adds a convention to internal code that we would have to remember.

**Comment by michalporeba** [comment]
There is much more than 'what will happen if there is no because missing'. It's just a quick illustration to the idea to further the discussion. That's why it is not a PR. As for the function name, I made it `Set-PesterResult` for now rather then `Set-TestResult` as I felt the latter is too generic, and since it will be a public method I'd like to limit potential for clashes with other modules that might be used. But I like the idea of calling it `Set-ItResult`, perhaps we could make it `Set-ItsResult` to make it more natural? Let me know your preference and I will include change. 

Few more questions
1. How about not displaying the error message for inconclusive results, make it more like the skipped example, with the 'because' description. I think it will be more useful to see why it is inconclusive, rather than the line that casued it to be inconclusive. We know what the test is so it is easy to find
2. I was thinking if there is no because provided I would add to the output a message suggesting that adding because might be a good idea 
3. Would adding `Set-ItResult -Pending` be useful to make it more complete? 
4. How about gutting the `Set-TestInconclusive` and call `Set-ItResult -Inconclusive` inside it while adding a message that `Set-TestInconclusive` is no deprecated and should be replaced with `Set-ItResult`? 
5. I get your point about internal conventions. I will change it to a switch. 

(I think there is a lot of refactoring and code cleanup that could be done in the code I have looked at, but I will not even attempt it right now ;) )
